### PR TITLE
Describe Navier-Stokes Diffeq example mesh by gmsh api usage and translate via FerriteGmsh

### PR DESCRIPTION
**Pull request by koehlerson**
Need to include an optional finer discretization with vortex street. Currently only coarse discretization for CI is in the literate file. Will test around when I'm back at my workstation

### CODE REVIEW

File: docs/src/literate/ns_vs_diffeq.jl [hunk 1]
```diff
@@ -121,38 +121,42 @@ using OrdinaryDiffEq
...
-    push!(hole_face_ring, FaceIndex((hci+x_cells, 1)))
-end
-grid.facesets["hole"] = Set(filter(x->x.idx[1] ∉ hole_cell_indices, collect(hole_face_ring)));
-cell_indices_map = map(ci->norm(mean(map(i->grid.nodes[i].x-[0.2,0.2], Ferrite.vertices(grid.cells[ci]))))>0.05 ? indexin([ci], cell_indices)[1] : 0, 1:length(grid.cells))
-grid.cells = grid.cells[cell_indices]
-for facesetname in keys(grid.facesets)
-    grid.facesets[facesetname] = Set(map(fi -> FaceIndex( cell_indices_map[fi.idx[1]] ,fi.idx[2]), collect(grid.facesets[facesetname])))
-end;
+# We specify first the rectangle, the cylinder, the surface spanned by the cylinder and the boolean difference of rectangle and cylinder.
+rect_tag = gmsh.model.occ.add_rectangle(0, 0, 0, 2.2, 0.41)
```
Comment (termi-official) [inline docs/src/literate/ns_vs_diffeq.jl:133]:
```suggestion
rect_tag = gmsh.model.occ.add_rectangle(0, 0, 0, 2.2, 0.41)
rect_tag = gmsh.model.occ.add_rectangle(0, 0, 0, 0.55, 0.41) #hide
```

File: docs/src/literate/ns_vs_diffeq.jl [hunk 2]
```diff
@@ -121,38 +121,42 @@ using OrdinaryDiffEq
...
+gmsh.model.model.add_physical_group(dim-1,[3],9,"right")
+gmsh.model.model.add_physical_group(dim-1,[1],10,"bottom")
+gmsh.model.model.add_physical_group(dim,[1],11,"domain")
+# Since we want a quad mesh, we specify the meshing algorithm to the quasi structured quad one.
+# For a complete list, [see the Gmsh docs](https://gmsh.info/doc/texinfo/gmsh.html#Mesh-options-list).
+gmsh.option.setNumber("Mesh.Algorithm",11)
+gmsh.option.setNumber("Mesh.MeshSizeFromCurvature",300)
+gmsh.option.setNumber("Mesh.MeshSizeMax",0.05)
+# reset the mesh size for the CI                        #src
+gmsh.option.setNumber("Mesh.MeshSizeFromCurvature",50)  #hide
```
Comment (termi-official) [inline docs/src/literate/ns_vs_diffeq.jl:153]:
```suggestion
gmsh.option.setNumber("Mesh.MeshSizeFromCurvature",10)  #hide
```

File: docs/src/literate/ns_vs_diffeq.jl (hunk 1 above)
Comment (termi-official) [reply docs/src/literate/ns_vs_diffeq.jl:133]: Note that in the CI I utilized a reduced domain (see below).

**Review by termi-official — CHANGES_REQUESTED** [review]
This might work.

**Comment by koehlerson** [comment]
@fredrikekre mesh from stokes example chokes :thinking:

**Comment by koehlerson** [comment]
I could not reproduce the vortex street with the updated mesh, but it's very likely not a problem with the mesh itself

**Comment by termi-official** [comment]
To elaborate on the likely issue: The solution for the formulation is not unique up to a constant value in the pressure. For the currently used mesh with equisized elements the linear solvers seems to not have trouble finding a close sequence of constants. We should fix this with the trick from the stokes example, where we force the constant to a proper value by setting the boundary integral of the pressure to zero. Does not seem to be super straight forward in this example.

**Comment by termi-official** [comment]
I could find several open issues.

1. the pressure ramp protocol causes discontinuities in the second derivative. The chosen solver uses a second derivative approximation to control the time step. This causes the failure at the end of the ramp.
2. Solver tolerances were too high for the unstructured grid to resolve the oscillations sufficiently.
3. Since we cannot enforce the time-dependent constraints consistently everywhere in OrdinaryDiffEq (interface does not allow this yet) the time step length degrades severly for the adaptive implicit Euler. Higher order methods just fail.
4. The evaluation of the right hand side is dead slow. After merging this we might want to open 2 issues here: One about speeding up the solver via [custom solver](https://docs.sciml.ai/DiffEqDocs/stable/features/linear_nonlinear/) and one about parallelizing the rhs evaluation (+user-defined, parallel Jacobian). The former should also be a first good step towards enforcing time-dependent constraints.
5. Internal norms were erroneous as they included the constraints.

I will push the fixes for 1,2,5 later today, after commenting everything. This should then make the PR ready to merge.

**Comment by termi-official** [comment]
Also as a future project idea: 3D Navier Stokes with MPI parallelization in tandem with OrdinaryDiffEq.

File: docs/src/literate/ns_vs_diffeq.jl [hunk 3]
```diff
@@ -403,7 +408,7 @@ function navierstokes!(du,u_uc,p,t)
         end
     end
 
-    # For now we have to ingore the evolution of the Dirichlet BCs.
+    # For now we have to ingore the evolution of the Dirichlet BCs. Note that these constraints
```
Comment (termi-official) [inline docs/src/literate/ns_vs_diffeq.jl:411]:
```suggestion
    # For now we have to ingore the evolution of the Dirichlet BCs.
```

File: docs/src/literate/ns_vs_diffeq.jl [hunk 4]
```diff
@@ -121,38 +117,45 @@ using OrdinaryDiffEq
 # We start off by defining our only material parameter.
 ν = 1.0/1000.0; #dynamic viscosity
 
-# Next a fine 2D rectangular grid has to be generated. We leave the cell size parametric for flexibility when
-# playing around with the code. Note that the mesh is pretty fine, leading to a high memory consumption when
+# Next a rectangular grid with a cylinder in it has to be generated.
+# We use `Gmsh` for the creation of the mesh and `FerriteGmsh` to translate it to a `Ferrite.Grid`
```
Comment (termi-official) [inline docs/src/literate/ns_vs_diffeq.jl:121]:
```suggestion
# We use `Gmsh` for the creation of the mesh and `FerriteGmsh` to translate it to a `Ferrite.Grid`.
```

File: docs/src/literate/ns_vs_diffeq.jl [hunk 5]
```diff
@@ -121,38 +117,45 @@ using OrdinaryDiffEq
...
+gmsh.option.setNumber("Mesh.MeshSizeFromCurvature",20)                                               #hide
+gmsh.option.setNumber("Mesh.MeshSizeMax",0.1)                                                        #hide
+# In the next step, the mesh is generated and finally translated.
+gmsh.model.mesh.generate(dim)
+grid = togrid()
+Gmsh.finalize()
 
 # We test against full development of the flow - so regenerate the grid                              #src
-grid = generate_grid(Quadrilateral, (x_cells, y_cells), Vec{2}((0.0, 0.0)), Vec{2}((0.55, 0.41)));   #hide
+grid = generate_grid(Quadrilateral, (55÷3, 41÷3), Vec{2}((0.0, 0.0)), Vec{2}((0.55, 0.41)));         #hide
```
Comment (termi-official) [inline docs/src/literate/ns_vs_diffeq.jl:158]:
```suggestion
```

File: docs/src/literate/ns_vs_diffeq.jl [hunk 6]
```diff
@@ -121,38 +117,45 @@ using OrdinaryDiffEq
...
+using FerriteGmsh
+using FerriteGmsh: Gmsh
+Gmsh.initialize()
+gmsh.option.set_number("General.Verbosity", 2)
+dim = 2;
+# We specify first the rectangle, the cylinder, the surface spanned by the cylinder and the boolean difference of rectangle and cylinder.
+rect_tag = gmsh.model.occ.add_rectangle(0, 0, 0, 2.2, 0.41)
+circle_tag = gmsh.model.occ.add_circle(0.2, 0.2, 0, 0.05)
+circle_curve_tag = gmsh.model.occ.add_curve_loop([circle_tag])
+circle_surf_tag = gmsh.model.occ.add_plane_surface([circle_curve_tag])
```
Comment (termi-official) [inline docs/src/literate/ns_vs_diffeq.jl:133]:
```suggestion
# We check for laminar flow development in the CI #src
if false #hide
rect_tag = gmsh.model.occ.add_rectangle(0, 0, 0, 2.2, 0.41)
circle_tag = gmsh.model.occ.add_circle(0.2, 0.2, 0, 0.05)
circle_curve_tag = gmsh.model.occ.add_curve_loop([circle_tag])
circle_surf_tag = gmsh.model.occ.add_plane_surface([circle_curve_tag])
else #hide
rect_tag = gmsh.model.occ.add_rectangle(0, 0, 0, 0.55, 0.41) #hide
end #hide
```

File: docs/src/literate/ns_vs_diffeq.jl [hunk 7]
```diff
@@ -121,38 +117,45 @@ using OrdinaryDiffEq
...
+gmsh.model.occ.cut([(dim,rect_tag)],[(dim,circle_surf_tag)]);
+# Now, the geometrical entities need to be synchronized in order to be available outside of `gmsh.model.occ`
+gmsh.model.occ.synchronize()
+# In the next lines, we add the physical groups needed to define boundary conditions.
+gmsh.model.model.add_physical_group(dim-1,[5],6,"hole")
```
Comment (termi-official) [inline docs/src/literate/ns_vs_diffeq.jl:138]:
```suggestion
if false #hide
gmsh.model.model.add_physical_group(dim-1,[5],6,"hole")
end #hide
```

**Review by termi-official — CHANGES_REQUESTED** [review]
Some final changes and this should be good to go.

File: docs/src/literate/ns_vs_diffeq.jl (hunk 7 above)
Comment (koehlerson) [reply docs/src/literate/ns_vs_diffeq.jl:138]: 10x dev trick

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 8]
```diff
@@ -1,3 +1,7 @@
+# Putting this flag to false reproduces the figure shown in the example #src
+# We check for laminar flow development in the CI                       #src
+IS_CI = true #hide
```
Comment (termi-official) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:3]: https://docs.julialang.org/en/v1/base/base/#Base.@isdefined
```suggestion
if !(@isdefined is_ci) #hide
    is_ci = false
end
```

Maybe something like this? @fredrikekre

Comment (koehlerson) [reply docs/src/literate-tutorials/ns_vs_diffeq.jl:3]: so we define `is_ci` in the beginning of the CI execution ?

Comment (termi-official) [reply docs/src/literate-tutorials/ns_vs_diffeq.jl:3]: https://github.com/Ferrite-FEM/Ferrite.jl/blob/master/docs/make.jl#L15C1-L15C44

Comment (koehlerson) [reply docs/src/literate-tutorials/ns_vs_diffeq.jl:3]: ah so I can change it already?

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 9]
```diff
@@ -1,3 +1,10 @@
+# Putting this flag to false reproduces the figure shown in the example #src
+# We check for laminar flow development in the CI                       #src
+if @isdefined is_ci    #hide
+    IS_CI = false      #hide
+else                   #hide
+    IS_CI = true       #hide
```
Comment (termi-official) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:6]: Is there a reason why we cannot use is_ci directly?

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 10]
```diff
@@ -180,21 +198,28 @@ ch = ConstraintHandler(dh);
...
 # is already enough to obtain some simple vortex streets. By increasing the
 # velocity further we can obtain stronger vortices - which may need additional
 # refinement of the grid.
 ∂Ω_inflow = getfacetset(grid, "left");
 
-vᵢₙ(t) = clamp(t, 0.0, 1.0)*1.0 #inflow velocity
-vᵢₙ(t) = clamp(t, 0.0, 1.0)*0.3 #hide
-parabolic_inflow_profile((x,y),t) = [4*vᵢₙ(t)*y*(0.41-y)/0.41^2,0]
+# !!! note
+#     The adaptivity only works if the pressure field varies smoothly
```
Comment (termi-official) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:216]:
```suggestion
#     Temporal adaptivity only works if the pressure field varies smoothly
```

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 11]
```diff
@@ -380,62 +440,138 @@ function navierstokes!(du,u_uc,p,t)
...
     progress=true, progress_steps=1,
-    saveat=Δt_save);
+    verbose=true, internalnorm=FreeDofErrorNorm(ch)
+);
+
 
 pvd = VTKFileCollection("vortex-street", grid);
-integrator = TimeChoiceIterator(integrator, 0.0:Δt_save:T)
-for (u_uc,t) in integrator
+for (u_uc,t) in TimeChoiceIterator(integrator, 0.0:Δt_save:T)
```
Comment (termi-official) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:574]:
```suggestion
for (u,t) in TimeChoiceIterator(integrator, 0.0:Δt_save:T)
```
And delete apply below. The solution should be computed correctly.

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 12]
```diff
@@ -467,7 +603,7 @@ function compute_divergence(dh, u, cellvalues_v)                            #hid
     return divv                                                             #hide
 end                                                                         #hide
 @testset "INS OrdinaryDiffEq" begin                                         #hide
```
Comment (termi-official) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:605]: We need to wrap this into `is_ci`.

File: docs/src/literate-tutorials/ns_vs_diffeq.jl (hunk 9 above)
Comment (koehlerson) [reply docs/src/literate-tutorials/ns_vs_diffeq.jl]: I think so, since `is_ci` is not defined if a user executes it locally

**Comment by koehlerson** [comment]
@oscardssmith  https://github.com/Ferrite-FEM/Ferrite.jl/pull/498/commits/e81ba5aaa156bebaed1c10d9d4c271de2270ae75#diff-86f6c62be033e4a97b5e334dd27ac6a9c3402556163dc345a26b5d49c1fa87d4R574-R587 `tuples(integrator)` looks correct, but the `TimeChoiceIterator` seems off

File: docs/src/literate-tutorials/ns_vs_diffeq.jl (hunk 9 above)
Comment (termi-official) [reply docs/src/literate-tutorials/ns_vs_diffeq.jl:6]: should this go to false?

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 13]
```diff
@@ -1,3 +1,10 @@
+# Putting this flag to false reproduces the figure shown in the example #src
+# We check for laminar flow development in the CI                       #src
+if @isdefined is_ci    #hide
+    IS_CI = is_ci      #hide
+else                   #hide
+    IS_CI = false      #hide
+end                    #hide
```
Comment (termi-official) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:7]: Turns out my proposal did not work out as I thought. Fredrik suggested the following fix

https://github.com/Ferrite-FEM/Ferrite.jl/blob/c33a191f618fa9dfb2d5b5efb3c6eae8000ba6d1/docs/src/literate-tutorials/reactive_surface.jl#L1-L8

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 14]
```diff
@@ -121,45 +125,62 @@ using OrdinaryDiffEq
...
+# of `gmsh.model.occ`
+gmsh.model.occ.synchronize()
+# In the next lines, we add the physical groups needed to define boundary conditions.
+if !IS_CI                                                                                           #hide
+bottomtag = gmsh.model.model.add_physical_group(dim-1,[6],-1,"bottom")
+lefttag = gmsh.model.model.add_physical_group(dim-1,[7],-1,"left")
+righttag = gmsh.model.model.add_physical_group(dim-1,[8],-1,"right")
+toptag = gmsh.model.model.add_physical_group(dim-1,[9],-1,"top")
+holetag = gmsh.model.model.add_physical_group(dim-1,[5],-1,"hole");
+nothing                                                                                             #hide
```
Comment (koehlerson) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:159]: Why is this still rendering a 10 output in the docs :cry:

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 15]
```diff
@@ -121,45 +125,62 @@ using OrdinaryDiffEq
...
+else                                                                                                #hide
+gmsh.model.model.add_physical_group(dim-1,[4],7,"left")                                             #hide
+gmsh.model.model.add_physical_group(dim-1,[3],8,"top")                                              #hide
+gmsh.model.model.add_physical_group(dim-1,[2],9,"right")                                            #hide
+gmsh.model.model.add_physical_group(dim-1,[1],10,"bottom");                                         #hide
+end # hide
```
Comment (termi-official) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:165]: xref ^
```suggestion
end # hide
nothing
```

**Comment by termi-official** [comment]
xref https://github.com/SciML/OrdinaryDiffEq.jl/issues/2238 the information needs to be added

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 16]
```diff
@@ -11,6 +19,9 @@
 #
 # ## Remarks on DifferentialEquations.jl
 #
+# !!! note "Required Version"
+#     This example will only work with OrdinaryDiffEq@v6.80.1.
```
Comment (koehlerson) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:23]:
```suggestion
#     This example will only work with OrdinaryDiffEq@v6.80.1. or above
```
@termi-official  ?

**Review by fredrikekre — APPROVED** [review]

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 17]
```diff
@@ -38,12 +49,8 @@
...
 # $\nu$ the dynamic viscosity and $\Delta$ the Laplacian. In the derivation we assumed
 # a constant density of 1 for the fluid and negligible coupling between the velocity components.
-# Finally we see that the pressure term appears only in combination with the gradient
-# operator, so for any solution $p$ the function $p + c$ is also an admissible solution, if
-# we do not impose Dirichlet conditions on the pressure. To resolve this we introduce the
-# implicit constraint that $ \int_\Omega p = 0 $.
 #
-# Our setup is derived from [Turek's DFG benchmark](http://www.mathematik.tu-dortmund.de/~featflow/en/benchmarks/cfdbenchmarking/flow/dfg_benchmark1_re20.html).
+# Our setup is derived from [Turek's DFG benchmark](http://www.mathematik.tu-dortmund.de/~featflow/en/benchmarks/cfdbenchmarking/flow/dfg_benchmark2_re100.html).
 # We model a channel with size $0.41 \times 2.2$ and a hole of radius $0.05$ centered at $(0.2, 0.2)$.
```
Comment (termi-official) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:54]: The description is now off as we changed the channel size at some point.
```suggestion
# We model a channel with size $0.41 \times 1.1$ and a hole of radius $0.05$ centered at $(0.2, 0.2)$.
```

File: docs/src/literate-tutorials/ns_vs_diffeq.jl [hunk 18]
```diff
@@ -1,6 +1,14 @@
+# Putting this flag to false reproduces the figure shown in the example #src
+# We check for laminar flow development in the CI                       #src
```
Comment (termi-official) [inline docs/src/literate-tutorials/ns_vs_diffeq.jl:2]: We should either update this comment or really check for the velocity profile in the test.

**Review by termi-official — APPROVED** [review]
LGTM otherwise.

File: docs/src/literate-tutorials/ns_vs_diffeq.jl (hunk 18 above)
Comment (koehlerson) [reply docs/src/literate-tutorials/ns_vs_diffeq.jl:2]:
```suggestion
# We check for a divergence free velocity field in the CI                      #src
```
?

Comment (termi-official) [reply docs/src/literate-tutorials/ns_vs_diffeq.jl:2]: That is fine for me.

**Pull request merged**